core/lib/contract_verifier/src/resolver/github/mod.rs

```diff
@@ -0,0 +1,279 @@
+            // the versions will be fetched later.
+            tracing::error!("failed syncing compiler versions at start: {:?}", err);
+        }
+
+        Ok(Self {
+            artifacts_dir,
+            gh_client,
+            client: reqwest::Client::new(),
+            supported_versions: RwLock::new(supported_versions),
+        })
+    }
+}
+
+impl GitHubCompilerResolver {
+    async fn download_version_if_needed(
+        &self,
+        compiler: CompilerType,
+        version: &str,
+    ) -> anyhow::Result<()> {
+        if compiler.exists(self.artifacts_dir.path(), version).await? {
+            tracing::debug!("Compiler {}:{} exists", compiler.as_str(), version);
+            return Ok(());
+        }
+
+        tracing::info!("Downloading {}:{}", compiler.as_str(), version);
+        let lock = self.supported_versions.read().await;
+        let versions = match compiler {
+            CompilerType::Solc => &lock.solc_versions,
+            CompilerType::ZkSolc => &lock.zksolc_versions,
+            CompilerType::Vyper => &lock.vyper_versions,
+            CompilerType::ZkVyper => &lock.zkvyper_versions,
+        };
+
+        let version_url = versions
+            .get(version)
+            .ok_or_else(|| {
+                ContractVerifierError::UnknownCompilerVersion("solc", version.to_owned())
+            })?
+            .clone();
+        drop(lock);
+        let path = compiler.bin_path_unchecked(self.artifacts_dir.path(), version);
+
+        let response = self.client.get(version_url).send().await?;
+        let body = response.bytes().await?;
+
+        tracing::info!("Saving {}:{} to {:?}", compiler.as_str(), version, path);
+
+        tokio::fs::create_dir_all(path.parent().unwrap())
+            .await
+            .map_err(|e| anyhow::anyhow!("failed to create dir: {}", e))?;
+
+        let mut file = tokio::fs::File::create(path)
+            .await
+            .map_err(|e| anyhow::anyhow!("failed to create file: {}", e))?;
+        file.write_all(&body)
+            .await
+            .map_err(|e| anyhow::anyhow!("failed to write to file: {}", e))?;
+        file.flush()
+            .await
+            .map_err(|e| anyhow::anyhow!("failed to flush file: {}", e))?;
+
+        // On UNIX-like systems, make file executable.
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            let mut perms = file.metadata().await?.permissions();
+            perms.set_mode(0o700); // Only owner can execute and access.
+            file.set_permissions(perms).await?;
+        }
+
+        tracing::info!("Finished downloading {}:{}", compiler.as_str(), version);
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl CompilerResolver for GitHubCompilerResolver {
+    async fn supported_versions(&self) -> anyhow::Result<SupportedCompilerVersions> {
+        let mut lock = self.supported_versions.write().await;
+        lock.update_if_needed(&self.gh_client).await?;
```

Comment:
I'm not sure if this method expected to return fast; the only consumer of this method is `ContractVerifier`, and it only calls it outside of the time-critical context. An additional task would further increase the cognitive complexity, so I intentionally decided against it.

It's possible to avoid locking for the whole duration of the update, even preserving non-atomic updates, but right now I see little value in that. Do you think it's a must?

Reply:
No, I'd consider it a "should have" though 🙃 Can be addressed in a follow-up.